Document the chat members plugin (#585)
grammyjs/website: site/docs/plugins/chat-members.md

# Chat Members Plugin (`chat-members`)

> Automatically store information about users on a chat and retrieve it easilly

Reviewer: ```suggestion
Automatically store information about users on a chat and retrieve it easily.
```

## Introduction

In many situations, it is necessary for a bot to have information about all the users of a given chat.
Currently, though, the Telegram Bot API exposes no method that allows us to retrieve this information.

> This plugin comes to the rescue: automatically listening to `chat_member` events and storing the `ChatMember` object.

Reviewer: ```suggestion
This plugin comes to the rescue: automatically listening to `chat_member` events and storing each `ChatMember` object.
```
or
```suggestion
This plugin comes to the rescue: automatically listening to `chat_member` events and storing the `ChatMember` objects.
```

> It then allows you to use `ctx.chatMembers.getChatMember()` to obtain information about a specific chat member, taking the previously stored information into account.

Reviewer: This is odd. First we say that there are no methods to keep track of the list. Then we give an example for retrieving info about a single member. As a reader, I'd ask myself why we can't just use `getChatMember` to do this, and what happened to the list of members.

Author: I'll remove this line.

> It then allows you to use `ctx.chatMembers.getChatMember()` to obtain information about a specific chat member, taking the previously stored information into account.

Author: ```suggestion
```

## Usage

### Storing Chat Members

You can use a valid grammY [storage adapter](https://grammy.dev/plugins/session.html#known-storage-adapters) or an
instance of any class that implements the [`StorageAdapter`](https://deno.land/x/grammy/mod.ts?s=StorageAdapter) interface.

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
import { Bot, Context, MemorySessionStorage } from "grammy";
import type { ChatMember } from "grammy/types";
import { chatMembers, ChatMembersFlavor } from "@grammyjs/chat-members";

type MyContext = Context & ChatMembersFlavor;

const adapter = new MemorySessionStorage<ChatMember>();

const bot = new Bot<MyContext>("<your bot token>");

bot.use(chatMembers(adapter));

bot.start({

>   allowed_updates: ["chat_member", "message"],

Reviewer: This needs some explanation. Why do we have this line?

>   onStart: ({ username }) => console.log(`Listening as ${username}`),

Reviewer: Unrelated to the example at hand.
```suggestion
```

});
```

</CodeGroupItem>

<CodeGroupItem title="JavaScript" active>

```js
import { Bot, Context, MemorySessionStorage } from "grammy";
import { chatMembers, ChatMembersFlavor } from "@grammyjs/chat-members";

const adapter = new MemorySessionStorage();

const bot = new Bot("<your bot token>");

bot.use(chatMembers(adapter));

bot.start({
  allowed_updates: ["chat_member", "message"],
  onStart: ({ username }) => console.log(`Listening as ${username}`),
});
```

</CodeGroupItem>

<CodeGroupItem title="Deno" active>

```ts
import {
  Bot,
  Context,
  MemorySessionStorage,
} from "https://deno.land/x/grammy/mod.ts";
import type { ChatMember } from "https://deno.land/x/grammy/types.ts";
import {
  chatMembers,
  ChatMembersFlavor,
} from "https://deno.land/x/grammy_chat_members/mod.ts";

type MyContext = Context & ChatMembersFlavor;

const adapter = new MemorySessionStorage<ChatMember>();

const bot = new Bot<MyContext>("<your bot token>");

bot.use(chatMembers(adapter));

bot.start({
  allowed_updates: ["chat_member", "message"],
  onStart: ({ username }) => console.log(`Listening as ${username}`),
});
```

</CodeGroupItem>

</CodeGroup>

### Reading Chat Members

> This plugin also adds a new `ctx.chatMembers.getChatMember` function that will check the storage for information about a chat member before querying Telegram for it.

Reviewer: As mentioned above. Shouldn't be presented as new information.

Alternatively, remove it from above.

If the chat member exists in the storage, it will be returned.
Otherwise, `ctx.api.getChatMember` will be called and the result will be saved to the storage, making subsequent calls faster and removing the need to call Telegram again for that user and chat in the future.

Here's an example:

```ts
bot.on("message", async (ctx) => {
  const chatMember = await ctx.chatMembers.getChatMember();

  return ctx.reply(
    `Hello, ${chatMember.user.first_name}! I see you are a ${chatMember.status} of this chat!`,
  );
});
```

> All parameters are optional; if you don't provide them, `ctx.chat.id` and `ctx.from.id` will be used instead.

Reviewer: Parameters of what? Perhaps we should have a second example illustrating these parameters.

> All parameters are optional; if you don't provide them, `ctx.chat.id` and `ctx.from.id` will be used instead.

Author: ```suggestion
This function accepts two parameters: the identifier for the chat, and the identifier for the user. You can use them like this:

```ts
bot.on("message", async (ctx) => {
  const chatMember = await ctx.chatMembers.getChatMember(ctx.chat.id, ctx.from.id);
  return ctx.reply(
    `Hello, ${chatMember.user.first_name}! I see you are a ${chatMember.status} of this chat!`,
  );
});
```

These parameters are optional. The default value for `chatId` is `ctx.chat.id`, and the default value for `userId` is `ctx.from.id`.

Reviewer: That nested list is an odd format, don't you think? 🤔

Author: I couldn't find a better way to make it concise, but clear 🤔

> Please notice that, if you don't provide a chat id and there's no `chat` property inside the context (for example, on inline query updates), this will throw an error.

Reviewer: ```suggestion
Please notice that, if you don't provide a chat identifier and there's no `chat` property inside the context (for example, on inline query updates), this will throw an error.
```

> Please notice that, if you don't provide a chat id and there's no `chat` property inside the context (for example, on inline query updates), this will throw an error.

Author: ```suggestion
Please notice that, if you don't provide a chat id and there's no `ctx.chat` property (for example, on inline query updates), this will throw an error.
```

> The same will happen if there's no `ctx.from` in the context.

Reviewer: This is inconsistent.

Use
```suggestion
The same will happen if there's no `from` in the context.
```
or change it to `ctx.chat` above, which I'd prefer.

Author: I'll change it above.

## Aggressive Storage

The `enableAggressiveStorage` config option will install middleware to cache chat members without depending on the `chat_member` event.
For every update, the middleware checks if `ctx.chat` and `ctx.from` exist.
If they both do, it then proceeds to call `ctx.chatMembers.getChatMember` to add the chat member information to the storage in case it doesn't exist.

> Please note that this means the storage will be called for **every update**, which may be a lot, depending on how many updates your bot receives.

Reviewer: ```suggestion
Please note that this means the storage will be written to for **every update**, which may be a lot, depending on how many updates your bot receives.
```

Author: It won't be written if the information is already there. It is only guaranteed to be read, to check if the record exists.

Author: Is this better?

```suggestion
Please note that this means the storage will be read for **every update**, which may be a lot, depending on how many updates your bot receives.
```

Reviewer: Maybe I am misunderstanding something here. I thought this:

Default:
- write to storage on chat_member
- read from storage as requested

Aggressive mode:
- write to storage for all updates that contain ctx.chat or ctx.from
- read from storage as requested

Is this accurate?

Author: Not really. Every time the plugin runs, it has to check to see if the current chat member already exists in the database, so it needs to do at least one read operation. If the member does not exist yet, then it does a write operation. By default, it only runs on `chat_member` updates. Aggressive mode makes it so it runs on every update, effectively doing a database read for every update.

Author: One thing we could do with the new storage adapter methods is keep an in-memory list of the known keys.
We could load all the keys on plugin setup and then always update both the in-memory list and the database whenever needed.

> Please note that this means the storage will be called for **every update**, which may be a lot, depending on how many updates your bot receives.

Author: Is this better?

```suggestion
Please note that this means the storage will be read for **every update**, which may be a lot, depending on how many updates your bot receives.
```

> This also has the potential to impact the performance of your bot drastically.

Reviewer: ```suggestion
This has the potential to impact the performance of your bot drastically.
```
We're still talking about the same thing.

> Only use this if you _really_ know what you're doing and are ok with the risks and consequences.

Reviewer: ```suggestion
Only use this if you _really_ know what you're doing and are okay with the risks and consequences.
```

## Plugin Summary

> - Name: `command-filter`

Reviewer: ```suggestion
- Name: `chat-members`
```

# Chat Members Plugin (`chat-members`)

> Automatically store information about users in a chat and retrieve it easily.

Reviewer: ```suggestion
Automatically store information about users in a chat and retrieve it easily.
Track group and channel members, and list them.
```

# Chat Members Plugin (`chat-members`)

> Automatically store information about users on a chat and retrieve it easily.

Reviewer: ```suggestion
Automatically store information about users in a chat and retrieve it easily.
```

## Introduction

In many situations, it is necessary for a bot to have information about all the users of a given chat.
Currently, though, the Telegram Bot API exposes no method that allows us to retrieve this information.

> This plugin comes to the rescue: automatically listening to `chat_member` events and storing each `ChatMember` objects.

Reviewer: ```suggestion
This plugin comes to the rescue: automatically listening to `chat_member` events and storing all `ChatMember` objects.
```

## Usage

### Storing Chat Members

> You can use a valid grammY [storage adapter](https://grammy.dev/plugins/session.html#known-storage-adapters) or an
> instance of any class that implements the [`StorageAdapter`](https://deno.land/x/grammy/mod.ts?s=StorageAdapter) interface.

Reviewer: 1S/L. Also, use relative links. Also also, link to MD files not HTML files.
```suggestion
You can use a valid grammY [storage adapter](./session.md#known-storage-adapters) or an instance of any class that implements the [`StorageAdapter`](https://deno.land/x/grammy/mod.ts?s=StorageAdapter) interface.
```

Reviewer 2: I badly missed these lmao

> Please note that, as per the official Telegram docs, your bot need to specify the `chat_member` update in the `allowed_updates` array, as shown in the example below.

Reviewer: ```suggestion
Please note that, as per the official Telegram docs, your bot needs to specify the `chat_member` update in the `allowed_updates` array, as shown in the example below.
```

> Please note that, as per the official Telegram docs, your bot needs to specify the `chat_member` update in the `allowed_updates` array, as shown in the example below.

Reviewer 3: ```suggestion
Please note that, as per the [official Telegram docs](https://core.telegram.org/bots/api#getupdates), your bot needs to specify the `chat_member` update in the `allowed_updates` array, as shown in the example below.
```

This means you also need to specify any other events you'd like to receive.

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts

> import { Bot, Context, MemorySessionStorage } from "grammy";

Reviewer: ```suggestion
import { Bot, type Context, MemorySessionStorage } from "grammy";
```

> import type { ChatMember } from "grammy/types";

Reviewer: Optionally, for consistency, do
```suggestion
import { type ChatMember } from "grammy/types";
```

> import { chatMembers, ChatMembersFlavor } from "@grammyjs/chat-members";

Reviewer: ```suggestion
import { chatMembers, type ChatMembersFlavor } from "@grammyjs/chat-members";
```

type MyContext = Context & ChatMembersFlavor;

const adapter = new MemorySessionStorage<ChatMember>();

const bot = new Bot<MyContext>("<your bot token>");

bot.use(chatMembers(adapter));

bot.start({
  // Make sure to specify the desired update types
  allowed_updates: ["chat_member", "message"],
});
```

</CodeGroupItem>

> <CodeGroupItem title="JavaScript" active>

Reviewer: ```suggestion
<CodeGroupItem title="JavaScript">
```

```js

> import { Bot, Context, MemorySessionStorage } from "grammy";
> import { chatMembers, ChatMembersFlavor } from "@grammyjs/chat-members";

Reviewer: ```suggestion
import { Bot, MemorySessionStorage } from "grammy";
import { chatMembers } from "@grammyjs/chat-members";
```

const adapter = new MemorySessionStorage();

const bot = new Bot("<your bot token>");

bot.use(chatMembers(adapter));

bot.start({
  // Make sure to specify the desired update types
  allowed_updates: ["chat_member", "message"],
});
```

</CodeGroupItem>

> <CodeGroupItem title="Deno" active>

Reviewer: ```suggestion
<CodeGroupItem title="Deno">
```

```ts
import {
  Bot,

>   Context,

Reviewer: ```suggestion
  type Context,
```

  MemorySessionStorage,
} from "https://deno.land/x/grammy/mod.ts";

> import type { ChatMember } from "https://deno.land/x/grammy/types.ts";

Reviewer: Optionally, for consistency, do
```suggestion
import { type ChatMember } from "https://deno.land/x/grammy/types.ts";
```

import {
  chatMembers,

>   ChatMembersFlavor,

Reviewer: ```suggestion
  type ChatMembersFlavor,
```

# Chat Members plugin (`chat-members`)

Automatically store information about users on a chat and retrieve it easilly

## Introduction

> In many situations, it is necessary for a bot to have information about all the users of a given chat. Currently, though, the Telegram Bot API exposes no method that allows that to happen.

Reviewer: ```suggestion
In many situations, it is necessary for a bot to have information about all the users of a given chat. Currently, though, the Telegram Bot API exposes no method that allows us to retrieve this information.
```

> This plugin does the next best thing: automatically listening to `chat_member` events and storing the `ChatMember` object. It then allows you to use `ctx.chatMembers.getChatMember()` to obtain information about a specific chat member, taking the previously stored information into account.

Reviewer: It's no worse. It's actually better, just less familiar.
```suggestion
This plugin comes to the rescue: automatically listening to `chat_member` events and storing the `ChatMember` object. It then allows you to use `ctx.chatMembers.getChatMember()` to obtain information about a specific chat member, taking the previously stored information into account.
```

## Usage

### Storing chat members

You can use a valid grammY [storage adapter](https://grammy.dev/plugins/session.html#known-storage-adapters) or an
instance of any class that implements the [`StorageAdapter`](https://deno.land/x/grammy/mod.ts?s=StorageAdapter)
interface.

> ```typescript

Reviewer: Needs to be included as TS/JS/Deno

Author: Right. Forgot about that. My bad.
I'll do it later today

Reviewer 2: I’ll do it for you.
